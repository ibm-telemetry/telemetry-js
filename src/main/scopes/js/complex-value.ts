/*
 * Copyright IBM Corp. 2023, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

/**
 * Object representing a complex value.
 */
export class ComplexValue {
  constructor(public readonly complexValue: unknown) {}
}
