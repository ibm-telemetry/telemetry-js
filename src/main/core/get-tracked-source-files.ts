/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { readFile } from 'node:fs/promises'
import * as path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from './log/logger.js'
import { TrackedFileEnumerator } from './tracked-file-enumerator.js'
import { parseDocument } from 'htmlparser2'
import type { Document as HtmlDocument } from 'domhandler'
import { HtmlParsedFile } from '../scopes/wc/interfaces.js'

/**
 * Gets all tracked source files to consider for data collection,
 * filtered by supplied file extension array.
 *
 * @param cwd - Working directory for underlying command execution.
 * @param root - Root directory in which to search for tracked source files. This is an absolute
 * path.
 * @param logger - Logger instance to use.
 * @param fileExtensions - List of file extensions to filter files by.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(
  cwd: string,
  root: string,
  logger: Logger,
  fileExtensions: string[]
) {
  logger.traceEnter('', 'getTrackedSourceFiles', [root, fileExtensions])

  const fileEnumerator = new TrackedFileEnumerator(logger)
  const files = []

  // If a file is passed instead of a directory, avoid the `git ls-tree` call
  if (fileExtensions.includes(path.extname(root))) {
    files.push(root)
  } else {
    files.push(
      ...(await fileEnumerator.find(cwd, root, (file) =>
        fileExtensions.includes(path.extname(file))
      ))
    )
  }

  const promises = files.map(async (file) => {
    return {
      fileName: file,
      async createSourceFile(): Promise<ts.SourceFile | HtmlDocument> {
        // Parse html file with `htmlparser2
        if (file.endsWith('html') || file.endsWith('htm')) {
          const parsedFile = parseDocument((await readFile(file)).toString()) as HtmlParsedFile
          parsedFile.fileName = file
        }

        // Parse any other files
        return ts.createSourceFile(
          file,
          (await readFile(file)).toString(),
          ts.ScriptTarget.ES2021,
          /* setParentNodes */ true
        )
      }
    }
  })

  const results = await Promise.all(promises)

  logger.traceExit(
    '',
    'getTrackedSourceFiles',
    results.map((result) => result.fileName)
  )
  return results
}
