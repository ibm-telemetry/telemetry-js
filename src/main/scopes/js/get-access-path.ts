/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import * as ts from 'typescript'

import type { Logger } from '../../core/log/logger.js'
import { ComplexValue } from './complex-value.js'
import { getNodeValueHandler } from './node-value-handler-map.js'

/**
 * Constructs a JsToken object from a given Identifier type AST node.
 *
 * @param node - TS node to retrieve access path for.
 * @param sourceFile - Root AST node.
 * @param logger - A logger instance.
 * @param currAccessPath - For internal use only, tracks current constructed access path.
 * @param topLevel - For internal use only, tracks top level function call.
 * @returns Access path to node represented as string array (chunks).
 */
export default function getAccessPath(
  node: ts.Node,
  sourceFile: ts.SourceFile,
  logger: Logger,
  currAccessPath: string[] = [],
  topLevel: boolean = true
): string[] {
  switch (node.kind) {
    case ts.SyntaxKind.Identifier:
      return [...currAccessPath, (node as ts.Identifier).escapedText.toString()]
    case ts.SyntaxKind.PropertyAccessExpression:
      currAccessPath.push((node as ts.PropertyAccessExpression).name.escapedText.toString())
      break
    case ts.SyntaxKind.ElementAccessExpression: {
      const argumentExpression = (node as ts.ElementAccessExpression).argumentExpression
      const data = getNodeValueHandler(argumentExpression.kind, sourceFile, logger).getData(
        argumentExpression
      )
      if (data instanceof ComplexValue) {
        currAccessPath = [
          ...currAccessPath,
          getAccessPath(argumentExpression, sourceFile, logger, [], false).reverse().join('.')
        ]
      } else {
        currAccessPath.push(data?.toString() ?? '')
      }
      break
    }
  }

  let accessPath = currAccessPath

  if ('expression' in node) {
    accessPath = getAccessPath(
      node.expression as ts.Node,
      sourceFile,
      logger,
      currAccessPath,
      false
    )
  }

  return topLevel ? accessPath.reverse() : accessPath
}
