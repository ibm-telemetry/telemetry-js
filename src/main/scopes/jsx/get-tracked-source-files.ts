/*
 * Copyright IBM Corp. 2023, 2023
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import path from 'node:path'

import * as ts from 'typescript'

import { type Logger } from '../../core/log/logger.js'
import { TrackedFileEnumerator } from '../../core/tracked-file-enumerator.js'

/**
 * Gets all tracked source files to consider for data collection.
 *
 * @param root - Root directory in which to search for tracked source files.
 * @param logger - Logger instance to use.
 * @returns An array of source file objects.
 */
export async function getTrackedSourceFiles(root: string, logger: Logger) {
  const fileEnumerator = new TrackedFileEnumerator(logger)
  const allowedSuffixes = ['js', 'jsx', 'ts', 'tsx']

  const files = await fileEnumerator.find(root, (file) =>
    allowedSuffixes.includes(path.extname(file))
  )

  const program = ts.createProgram(files, {})

  return program.getSourceFiles().filter((file) => !file.isDeclarationFile)
}
