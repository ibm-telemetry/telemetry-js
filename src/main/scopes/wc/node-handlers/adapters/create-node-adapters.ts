/*
 * Copyright IBM Corp. 2025, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { type Node as HtmlNode } from 'domhandler'
import type * as ts from 'typescript'

import type { INodeAdapter, ParsedFile } from '../../interfaces.js'
import { HtmlNodeAdapter } from './html-node-adapter.js'
import { isHtmlDocument } from '../../utils/is-html-document.js'
import { isTsSourceFile } from '../../utils/is-ts-source-file.js'
import { TsNodeAdapter } from './ts-node-adapter.js'

/**
 *
 * @param sourceFile
 */
export function createNodeAdapter(sourceFile: ParsedFile): INodeAdapter {
  if (isTsSourceFile(sourceFile)) {
    return new TsNodeAdapter(sourceFile as ts.Node)
  }

  if (isHtmlDocument(sourceFile)) {
    return new HtmlNodeAdapter(sourceFile as HtmlNode)
  }

  throw new Error('Unsupported file type')
}
