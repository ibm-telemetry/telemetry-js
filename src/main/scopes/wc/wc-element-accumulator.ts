/*
 * Copyright IBM Corp. 2025, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { JsxElement } from 'typescript'

import type { JsImport } from '../js/interfaces.js'
import { JsAccumulator } from '../js/js-accumulator.js'
import type { CdnImport, WcElement } from './interfaces.js'

/**
 * Responsible for maintaining an aggregated state of imports and elements.
 */
export class WcElementAccumulator extends JsAccumulator {
  public cdnImports: CdnImport[]
  public readonly elements: (WcElement | JsxElement)[]
  public readonly elementImports: Map<WcElement | JsxElement, JsImport | CdnImport>
  public readonly scriptSources: string[]

  constructor() {
    super()
    this.cdnImports = []
    this.elements = []
    this.elementImports = new Map()
    this.scriptSources = []
  }
}
