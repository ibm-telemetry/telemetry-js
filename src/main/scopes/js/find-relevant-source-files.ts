/*
 * Copyright IBM Corp. 2024, 2024
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import getPropertyByPath = require('lodash/get.js')
import * as path from 'node:path'

import { ObjectPath } from 'object-scan'

import { getTrackedSourceFiles } from '../../core/get-tracked-source-files.js'
import { Logger } from '../../core/log/logger.js'
import { NoInstallationFoundError } from '../../exceptions/no-installation-found-error.js'
import { getDependencyTree } from '../npm/get-dependency-tree.js'
import { getDirectoryPrefix } from '../npm/get-directory-prefix.js'
import { getInstalledVersionPaths } from '../npm/get-installed-version-paths.js'
import { getPackageData } from '../npm/get-package-data.js'
import { getPackageTrees } from '../npm/get-package-trees.js'
import { getTreePredecessor } from '../npm/get-tree-predecessor.js'
import { DependencyTree, PackageData } from '../npm/interfaces.js'

/**
 * Finds tracked source files and then filters them based on ones that appear in a  project which
 * depends on the in-context instrumented package/version.
 *
 * @param instrumentedPackage - Data about the instrumented package to use during filtering.
 * @param cwd - Current working directory. This must be inside of the root directory. This is an
 * absolute path.
 * @param root - Root-most directory. This is an absolute path.
 * @param fileExtensions - List of file extensions to capture metrics for.
 * @param logger - Logger instance.
 * @returns A (possibly empty) array of source files.
 */
export async function findRelevantSourceFiles(
  instrumentedPackage: PackageData,
  cwd: string,
  root: string,
  fileExtensions: string[],
  logger: Logger
) {
  logger.traceEnter('', 'findRelevantSourceFiles', [instrumentedPackage, cwd, root, fileExtensions])
  const sourceFiles = await getTrackedSourceFiles(root, logger, fileExtensions)

  const dependencyTree = await getDependencyTree(cwd, root, logger)

  const filterPromises = sourceFiles.map(async (f) => {
    const prefix = await getDirectoryPrefix(path.dirname(f.fileName), logger)
    const prefixPackageData = await getPackageData(prefix, root, logger)

    let packageTrees = getPackageTrees(dependencyTree, prefixPackageData)

    let instrumentedInstallVersions: string[] | undefined = undefined
    let shortestPathLength: number | undefined = undefined
    do {
      for (const tree of packageTrees) {
        const instrumentedInstallPaths = getInstalledVersionPaths(tree, instrumentedPackage.name)
        if (instrumentedInstallPaths.length > 0) {
          const pathsLength = instrumentedInstallPaths[0]?.length ?? 0
          if (shortestPathLength === undefined || pathsLength < shortestPathLength) {
            instrumentedInstallVersions = instrumentedInstallPaths.map(
              (path) => getPropertyByPath(tree, path)['version']
            )
            shortestPathLength = pathsLength
          }
        }
      }
      // did not find, go up one level for all packages
      packageTrees = packageTrees
        .map((tree) => getTreePredecessor(dependencyTree, tree['path'] as ObjectPath))
        .filter((tree) => tree !== undefined) as DependencyTree[]
    } while (shortestPathLength === undefined && packageTrees.length > 0)

    if (instrumentedInstallVersions === undefined) {
      throw new NoInstallationFoundError(instrumentedPackage.name)
    }

    return instrumentedInstallVersions.some((version) => version === instrumentedPackage.version)
  })
  const filterData = await Promise.all(filterPromises)

  const results = sourceFiles.filter((_, index) => {
    return filterData[index]
  })

  logger.traceEnter('', 'findRelevantSourceFiles', results)
  return results
}
