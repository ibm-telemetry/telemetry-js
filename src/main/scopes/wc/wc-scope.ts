/*
 * Copyright IBM Corp. 2023, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type * as ts from 'typescript'

import { Trace } from '../../core/log/trace.js'
import { Scope } from '../../core/scope.js'
import { EmptyScopeError } from '../../exceptions/empty-scope.error.js'
import { findRelevantSourceFiles } from '../js/find-relevant-source-files.js'
import type { JsImportMatcher } from '../js/interfaces.js'
import { processFile } from '../js/process-file.js'
import { removeIrrelevantImports } from '../js/remove-irrelevant-imports.js'
import { JsxElementAllImportMatcher } from '../jsx/import-matchers/jsx-element-all-import-matcher.js'
import { JsxElementNamedImportMatcher } from '../jsx/import-matchers/jsx-element-named-import-matcher.js'
import { JsxElementRenamedImportMatcher } from '../jsx/import-matchers/jsx-element-renamed-import-matcher.js'
// import { type WcElement } from './interfaces.js'
import { type JsxElement } from '../jsx/interfaces.js'
import { ElementMetric } from '../jsx/metrics/element-metric.js'
import { getPackageData } from '../npm/get-package-data.js'
import type { PackageData } from '../npm/interfaces.js'
import { WcElementAccumulator } from './wc-element-accumulator.js'
import { jsxNodeHandlerMap } from './wc-node-handler-map.js'

/**
 * Scope class dedicated to data collection from a DOM-based environment.
 */
export class WcScope extends Scope {
  public static readonly fileExtensions = [
    '.js',
    '.mjs',
    '.cjs',
    '.jsx',
    '.tsx',
    '.mjsx',
    '.cjsx',
    '.mtsx',
    '.ctsx',
    '.html',
    '.htm'
  ]

  public override name = 'wc' as const
  private runSync = false

  /**
   * Entry point for the scope. All scopes run asynchronously.
   */
  @Trace()
  public override async run(): Promise<void> {
    const collectorKeys = this.config.collect[this.name]
    if (collectorKeys === undefined || Object.keys(collectorKeys).length === 0) {
      throw new EmptyScopeError(this.name)
    }

    const promises: Array<Promise<void>> = []

    Object.keys(collectorKeys).forEach((key) => {
      switch (key) {
        case 'elements':
          promises.push(this.captureAllMetrics())
          break
      }
    })

    await Promise.allSettled(promises)
  }

  /**
   * Generates metrics for all discovered instrumented jsx elements found in the current working
   * directory's project.
   */
  @Trace()
  async captureAllMetrics(): Promise<void> {
    const importMatchers = [
      new JsxElementAllImportMatcher(),
      new JsxElementNamedImportMatcher(),
      new JsxElementRenamedImportMatcher()
    ]
    const instrumentedPackage = await getPackageData(this.cwd, this.cwd, this.logger)
    const sourceFiles = await findRelevantSourceFiles(
      instrumentedPackage,
      this.cwd,
      this.root,
      WcScope.fileExtensions,
      this.logger
    )

    this.logger.debug('Filtered source files: ' + sourceFiles.map((f) => f.fileName))

    const promises: Promise<void>[] = []

    for (const sourceFile of sourceFiles) {
      // route JS/JSX differently

      if (sourceFile.fileName.endsWith('html')) {
      } else {
      }

      // if JS/JSX
      const resultPromise = this.captureFileMetrics(
        await sourceFile.createSourceFile(),
        instrumentedPackage,
        importMatchers
      )

      // if html
      //

      if (this.runSync) {
        await resultPromise
      } else {
        promises.push(resultPromise)
      }
    }

    await Promise.allSettled(promises)
  }

  /**
   * Generates metrics for all discovered instrumented jsx elements found
   * in the supplied SourceFile node.
   *
   * @param sourceFile - The sourcefile node to generate metrics for.
   * @param instrumentedPackage - Name and version of the instrumented package
   * to capture metrics for.
   * @param importMatchers - Matchers instances to use for import-element matching.
   */
  async captureFileMetrics(
    sourceFile: ts.SourceFile,
    instrumentedPackage: PackageData,
    importMatchers: JsImportMatcher<JsxElement>[]
  ) {
    const accumulator = new WcElementAccumulator()

    // TODO make a HTML node handler
    processFile(accumulator, sourceFile, jsxNodeHandlerMap, this.logger)

    this.logger.debug('Pre0filter accumulator contents:', JSON.stringify(accumulator))

    removeIrrelevantImports(accumulator, instrumentedPackage.name)
    this.resolveElementImports(accumulator, importMatchers)

    accumulator.elements.forEach((jsxElement) => {
      const jsImport = accumulator.elementImports.get(jsxElement)

      if (jsImport === undefined) {
        return
      }

      this.capture(
        new ElementMetric(jsxElement, jsImport, instrumentedPackage, this.config, this.logger)
      )
    })
  }

  resolveElementImports(
    accumulator: WcElementAccumulator,
    elementMatchers: JsImportMatcher<JsxElement>[]
  ) {
    accumulator.elements.forEach((jsxElement) => {
      const jsImport = elementMatchers
        .map((elementMatcher) => elementMatcher.findMatch(jsxElement, accumulator.imports))
        .find((jsImport) => jsImport !== undefined)

      if (jsImport === undefined) {
        return
      }

      accumulator.elementImports.set(jsxElement, jsImport)
    })
  }

  /**
   * **For testing purposes only.**
   * Makes the JsxScope collection run "synchronously" (one source file at a time). Defaults to
   * `false`.
   *
   * @param runSync - Boolean of whether or not to run synchronously.
   */
  setRunSync(runSync: boolean) {
    this.runSync = runSync
  }
}
