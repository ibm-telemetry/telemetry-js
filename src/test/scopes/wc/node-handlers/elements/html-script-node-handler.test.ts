/*
 * Copyright IBM Corp. 2025, 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import type { Node as HtmlNode } from 'domhandler'
import { describe, expect, it } from 'vitest'

import { getTrackedSourceFiles } from '../../../../../main/core/get-tracked-source-files.js'
import type { HtmlParsedFile } from '../../../../../main/scopes/wc/interfaces.js'
import type { INodeAdapter } from '../../../../../main/scopes/wc/interfaces.js'
import { createNodeAdapter } from '../../../../../main/scopes/wc/node-handlers/adapters/create-node-adapters.js'
import { WcScriptNodeHandler } from '../../../../../main/scopes/wc/node-handlers/elements/html-script-node-handler.js'
import { WcElementAccumulator } from '../../../../../main/scopes/wc/wc-element-accumulator.js'
import { WcScope } from '../../../../../main/scopes/wc/wc-scope.js'
import { findNodesByType } from '../../../../__utils/find-nodes-by-type.js'
import { Fixture } from '../../../../__utils/fixture.js'
import { initLogger } from '../../../../__utils/init-logger.js'

describe('class: WcScriptNodeHandler', async () => {
  const logger = initLogger()
  const fixture = new Fixture('projects/web-components-project/test.html')
  const sourceFile = (await (
    await getTrackedSourceFiles(fixture.path, fixture.path, logger, WcScope.fileExtensions)
  )[0]?.createSourceFile()) as HtmlParsedFile
  const accumulator = new WcElementAccumulator()
  const handler = new WcScriptNodeHandler(sourceFile, logger)
  it('correctly returns the script sources for an html file fixture', async () => {
    const rootAdapter = createNodeAdapter(sourceFile) as INodeAdapter
    //@ts-expect-error typescript not accepting adapter for type ts source file
    const wcScriptSourceNodes = findNodesByType(rootAdapter, 'HtmlScript') as HtmlNode[]

    wcScriptSourceNodes.forEach((scriptSource) => {
      handler.handle(scriptSource, accumulator)
    })

    expect(accumulator.scriptSources).toMatchSnapshot()
  })
})
