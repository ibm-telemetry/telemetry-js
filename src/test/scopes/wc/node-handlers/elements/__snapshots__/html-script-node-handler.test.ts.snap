// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`class: WcScriptNodeHandler > correctly returns the script sources for an html file fixture 1`] = `
[
  "./js-file-with-import.js",
]
`;
